File: java/Solution77.java
package leetCode.java;

import java.util.*;

//Backtracking O(n * n!) O(n)
class Solution77 {
    public List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combine(int n, int k) {
        backtracking(1, new ArrayList<>(), n, k);
        return res;
    }

    public void backtracking(int start, List<Integer> nums, int n, int k){
        if(nums.size() == k){
            res.add(new ArrayList(nums));
            return;
        }

        for(int i=start; i<=n; ++i){
            nums.add(i);
            backtracking(i + 1, nums, n, k);
            nums.remove(nums.size() - 1);
        }
    }
}

File: c++/Solution77.cpp
//Backtracking O(n * n!) O(n)
class Solution77 {
public:
    vector<vector<int>> res;
    
    vector<vector<int>> combine(int n, int k) {
        vector<int> selected;
        backtracking(1, selected, n, k);
        return res;
    }

    void backtracking(int start, vector<int>& selected, int n, int k) {
        if (selected.size() == k) {
            res.push_back(selected);
        }

        for (int i = start; i <= n; ++i) {
            selected.push_back(i);
            backtracking(i + 1, selected, n, k);
            selected.pop_back();
        }
    }
};